src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -150,8 +163,30 @@ private void AddRange(IEnumerable<KeyValuePair<TKey, TValue>> collection)
                 return;
             }
 
-            // Fallback path for IEnumerable that isn't a non-subclassed Dictionary<TKey,TValue>.
-            foreach (KeyValuePair<TKey, TValue> pair in collection)
+            // We similarly special-case KVP<>[] and List<KVP<>>, as they're commonly used to seed dictionaries, and
+            // we want to avoid the enumerator costs (e.g. allocation) for them as well.
+
+            if (enumerable is KeyValuePair<TKey, TValue>[] array)
+            {
+                foreach (KeyValuePair<TKey, TValue> pair in array)
+                {
+                    Add(pair.Key, pair.Value);
+                }
+                return;
+            }
+
+            if (enumerable.GetType() == typeof(List<KeyValuePair<TKey, TValue>>))
+            {
+                foreach (KeyValuePair<TKey, TValue> pair in CollectionsMarshal.AsSpan((List<KeyValuePair<TKey, TValue>>)enumerable))
+                {
+                    Add(pair.Key, pair.Value);
+                }
+                return;
+            }
```

Comment:
Maybe make a helper local method for getting a `ReadOnlySpan` to deduplicate the loop here?